modules/ml/src/svmsgd.cpp

```diff
@@ -0,0 +1,517 @@
+}
+
+void SVMSGDImpl::writeParams( FileStorage& fs ) const
+{
+    String SvmsgdTypeStr;
+
+    switch (params.svmsgdType)
+    {
+    case SGD:
+        SvmsgdTypeStr = "SGD";
+        break;
+    case ASGD:
+        SvmsgdTypeStr = "ASGD";
+        break;
+    default:
+        SvmsgdTypeStr = format("Unknown_%d", params.svmsgdType);
+    }
+
+    fs << "svmsgdType" << SvmsgdTypeStr;
+
+    String marginTypeStr;
+
+    switch (params.marginType)
+    {
+    case SOFT_MARGIN:
+        marginTypeStr = "SOFT_MARGIN";
+        break;
+    case HARD_MARGIN:
+        marginTypeStr = "HARD_MARGIN";
+        break;
+    default:
+        marginTypeStr = format("Unknown_%d", params.marginType);
+    }
+
+    fs << "marginType" << marginTypeStr;
+
+    fs << "marginRegularization" << params.marginRegularization;
+    fs << "initialStepSize" << params.initialStepSize;
+    fs << "stepDecreasingPower" << params.stepDecreasingPower;
+
+    fs << "term_criteria" << "{:";
+    if( params.termCrit.type & TermCriteria::EPS )
+        fs << "epsilon" << params.termCrit.epsilon;
+    if( params.termCrit.type & TermCriteria::COUNT )
+        fs << "iterations" << params.termCrit.maxCount;
+    fs << "}";
+}
+
+void SVMSGDImpl::read(const FileNode& fn)
+{
+    clear();
+
+    readParams(fn);
+
+    fn["weights"] >> weights_;
+    fn["shift"] >> shift_;
+}
+
+void SVMSGDImpl::readParams( const FileNode& fn )
+{
+    String svmsgdTypeStr = (String)fn["svmsgdType"];
+    int svmsgdType =
+            svmsgdTypeStr == "SGD" ? SGD :
+                                     svmsgdTypeStr == "ASGD" ? ASGD : -1;
+
+    if( svmsgdType < 0 )
+        CV_Error( CV_StsParseError, "Missing or invalid SVMSGD type" );
+
+    params.svmsgdType = svmsgdType;
+
+    String marginTypeStr = (String)fn["marginType"];
+    int marginType =
+            marginTypeStr == "SOFT_MARGIN" ? SOFT_MARGIN :
+                                     marginTypeStr == "HARD_MARGIN" ? HARD_MARGIN : -1;
+
+    if( marginType < 0 )
+        CV_Error( CV_StsParseError, "Missing or invalid margin type" );
+
+    params.marginType = marginType;
+
+    CV_Assert ( fn["marginRegularization"].isReal() );
+    params.marginRegularization = (float)fn["marginRegularization"];
+
+    CV_Assert ( fn["initialStepSize"].isReal() );
+    params.initialStepSize = (float)fn["initialStepSize"];
+
+    CV_Assert ( fn["stepDecreasingPower"].isReal() );
+    params.stepDecreasingPower = (float)fn["stepDecreasingPower"];
+
+    FileNode tcnode = fn["term_criteria"];
+    if( !tcnode.empty() )
+    {
+        params.termCrit.epsilon = (double)tcnode["epsilon"];
+        params.termCrit.maxCount = (int)tcnode["iterations"];
+        params.termCrit.type = (params.termCrit.epsilon > 0 ? TermCriteria::EPS : 0) +
+                (params.termCrit.maxCount > 0 ? TermCriteria::COUNT : 0);
+    }
+    else
+        params.termCrit = TermCriteria( TermCriteria::EPS + TermCriteria::COUNT, 100000, FLT_EPSILON );
```

Comment:
I think a better idea is to call `setOptimalParameters` first. And do nothing if some of the parameters weren't set.